WalletWasabi/Services/WalletService.cs

```diff
@@ -630,6 +638,148 @@ public async Task<Block> GetOrDownloadBlockAsync(uint256 hash, CancellationToken
 					cancel.ThrowIfCancellationRequested();
 					try
 					{
+						// Try to get block information from local running Core node first.
+						try
+						{
+							using (var localNode = Node.ConnectToLocal(Network))
```

Comment:
I think there is an alternative approach to achieve this without this code. IMO with should try to connect to a local node before trying to connect to the p2p network because if we have our own a trusted local node then it is unnecessary to connect to somebody else's random nodes. Doing that, this code is not necessary any more. Moreover, we can save a lot of bandwidth too what could be something very good in the future.

Reply:
This PR is not Bitcoin Core integration. This PR is an optimization on block sources. If we already have the block, then let's not ask the internet about it.

This code we checks if Core is running on-demand. Core has way too much configuration we cannot rely on it. For example pruned nodes wouldn't be able to serve historic blocks and that'd mean people get stuck at recovery. Or non-listening nodes, which is 80% of the network doesn't work with this PR, for that we'd need RPC. Mempool is also not included in this PR, nor we could rely on only on local Bitcoin Core mempool properly, since there are many configurations those can ruin the flow of reliable mempool info. Also what if the user shuts down Core or starts it again? What you are talking about is bringing up a host of other questions we don't have to care about right now, if the integration is on-demand.

I don't think we will integrate Core anytime soon, that's way too much work and maintenance, but fetching blocks from available sources is just a low hanging fruit.